llvm/lib/MCCAS/MCCASObjectV1.cpp

```diff
@@ -3087,12 +3087,29 @@ mccasformats::v1::loadDIETopLevel(DIETopLevelRef TopLevelRef) {
 }
 
 struct DIEVisitor {
+
+  struct AbbrevDIEContents {
+    dwarf::Attribute Attr;
+    dwarf::Form Form;
+    bool FormInDistinctData;
+    std::optional<uint8_t> FormSize;
+  };
+
+  struct AbbrevDIETagAndChildren {
+    dwarf::Tag Tag;
+    bool HasChildren;
+  };
+
   Error visitDIERef(DIEDataRef Ref);
   Error visitDIERef(BinaryStreamReader &DataReader, unsigned AbbrevIdx,
                     StringRef DIEData, ArrayRef<DIEDataRef> &DIEChildrenStack);
-  Error visitDIEAttrs(AbbrevEntryReader &AbbrevReader,
-                      BinaryStreamReader &Reader, StringRef DIEData);
+  Error visitDIEAttrs(BinaryStreamReader &Reader, StringRef DIEData,
+                      unsigned AbbrevIdx);
+  Error materializeAbbrevDIE(unsigned AbbrevIdx);
 
+  SmallVector<
+      std::pair<AbbrevDIETagAndChildren, SmallVector<AbbrevDIEContents>>>
```

Comment:
Reading how this thing is used in subsequent commits, I think we can simplify the code a lot if we do away with the pair and fold it into `AbbrevDIETagAndChildren`

```
struct AbbrevEntry {
    dwarf::Tag Tag;
    bool HasChildren;
    SmallVector<AbbrevContent>  AbbrevContents;
};
```

(btw the "DIE" wording feels off and we could strip it?)

Reply:
Sounds plausible to me. @rastogishubham ?